### PR TITLE
Encode BPF instructions to mitigate heap spray attacks

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Encode BPF instructions to mitigate heap spray attacks.

Normally the [no-execute bit](https://en.wikipedia.org/wiki/NX_bit) on memory pages prevents an attacker from injecting shell code via heap spray and triggering a jump into executing it. The presence of an interpreter in the process nullifies the protection offered by the no-execute bit on code pages as the attacker can trigger a jump to the start of the interpreter to execute the injected byte code.

This change XOR's BPF byte code with the base address of the memory holding the byte code, which prevents trivial heap-spray attacks from working. In addition it provides an option to add in a per-process secret to further raise the difficulty.

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "files.associations": {
+        "inttypes.h": "c"
+    }
+}
```
Comment (matt-gretton-dann) [inline .vscode/settings.json:5]: This looks like an addition related to choice of IDE/Editor.  Probably doesn't belong in this PR.

(I'm not adverse to adding some VSCode setup support, but that should be a separate PR I think).

**Review by matt-gretton-dann — COMMENTED** [review]
I like the general idea.  My concern is whether it is possible to change the secret in use by calls to `ubpf_set_pointer_secret()` after `ubpf_store_instruction()`, and accidentally stop being able to run code.

Maybe if the secret was a member of the `ubpf_vm` struct we can then ensure that `ubpf_set_pointer_secret()` is not called after `ubpf_store_instruction()` has been called on a VM.  Additionally this would allow different secrets for different VMs.

File: .vscode/settings.json (hunk 1 above)
Comment (Alan-Jowett) [reply .vscode/settings.json:5]: Oops. Added by mistake. Let me take that out.

**Comment by Alan-Jowett** [comment]
Modified it to store secret on a per-VM basis and only permit setting secret if there are no instructions loaded.

**Review by matt-gretton-dann — APPROVED** [review]
LGTM.
